### PR TITLE
[16.0][ADD] edi_notification_oca

### PR DESCRIPTION
**Pull request by QuocDuong1306**

### CODE REVIEW

File: edi_notification_oca/models/edi_exchange_type.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+
+from odoo import _, api, fields, models
+from odoo.exceptions import ValidationError
+
+
+class EDIExchangeType(models.Model):
+    _inherit = "edi.exchange.type"
+
+    send_notification_on_error = fields.Boolean(
```
Comment (simahawk) [inline edi_notification_oca/models/edi_exchange_type.py]:
```suggestion
    notify_on_process_error = fields.Boolean(
```
this is just for `process` action

File: edi_notification_oca/models/edi_exchange_type.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+            self.notification_users_ids = None
+
+    @api.constrains(
+        "send_notification_on_error",
+        "notification_groups_ids",
+        "notification_users_ids",
+    )
+    def _check_notification_on_error_groups_users(self):
+        for rec in self:
+            if not rec.send_notification_on_error:
```
Comment (simahawk) [inline edi_notification_oca/models/edi_exchange_type.py]: I'm not sure this constraint is useful. The other fields should be hidden when the flag is off and you already have the onchange to wipe the fields.

File: edi_notification_oca/readme/CONTRIBUTORS.rst [hunk 3]
```diff
@@ -0,0 +1 @@
+* Duong (Tran Quoc) <duontq@troz.com>
```
Comment (simahawk) [inline edi_notification_oca/readme/CONTRIBUTORS.rst:1]:
```suggestion
* Duong (Tran Quoc) <duontq@troz.com>
* Simone Orsi <simone.orsi@camptocamp.com>
```

File: edi_notification_oca/tests/test_edi_notification.py [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+        cls.user_a = cls._create_user(cls, "A")
+        cls.user_b = cls._create_user(cls, "B")
+        cls.user_c = cls._create_user(cls, "C")
+
+    def setUp(self):
+        super().setUp()
+        FakeInputProcess.reset_faked()
+
+    def _create_user(self, letter):
+        return self.env["res.users"].create(
```
Comment (simahawk) [inline edi_notification_oca/tests/test_edi_notification.py]:
```suggestion
        return self.env["res.users"].with_context(no_reset_password=True).create(
```
Always :wink: 
Or... use `new_test_user` from `odoo.tests.common`.

File: edi_notification_oca/models/edi_exchange_type.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+class EDIExchangeType(models.Model):
+    _inherit = "edi.exchange.type"
+
+    send_notification_on_error = fields.Boolean(
+        help="If an error happens on process, a notification will be sent to all"
+        " selected users. If active, please select the specific groups and"
+        " specific users in the 'Notifications' page.",
+        default=False,
+    )
+    notification_groups_ids = fields.Many2many(
```
Comment (simahawk) [inline edi_notification_oca/models/edi_exchange_type.py]:
```suggestion
    notify_on_process_error_groups_ids = fields.Many2many(
```

File: edi_notification_oca/models/edi_exchange_type.py [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+        comodel_name="res.groups", string="Notification Groups"
+    )
+    notification_users_ids = fields.Many2many(
```
Comment (simahawk) [inline edi_notification_oca/models/edi_exchange_type.py]:
```suggestion
    notify_on_process_error_users_ids = fields.Many2many(
```

File: edi_notification_oca/models/edi_exchange_type.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    _inherit = "edi.exchange.type"
+
+    send_notification_on_error = fields.Boolean(
+        help="If an error happens on process, a notification will be sent to all"
+        " selected users. If active, please select the specific groups and"
+        " specific users in the 'Notifications' page.",
+        default=False,
+    )
+    notification_groups_ids = fields.Many2many(
+        comodel_name="res.groups", string="Notification Groups"
```
Comment (simahawk) [inline edi_notification_oca/models/edi_exchange_type.py]:
```suggestion
        comodel_name="res.groups", string="Notify groups on process error"
```

File: edi_notification_oca/models/edi_exchange_type.py [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+    )
+    notification_users_ids = fields.Many2many(
+        comodel_name="res.users",
+        string="Notification Users",
```
Comment (simahawk) [inline edi_notification_oca/models/edi_exchange_type.py]:
```suggestion
        string="Notify users on process error",
```

File: edi_notification_oca/components/listener.py [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+        users = self.env["res.users"]
+        if exc_type.notification_groups_ids:
+            for group in exc_type.notification_groups_ids:
+                users |= group.users
+        if exc_type.notification_users_ids:
+            users |= exc_type.notification_users_ids
+        # Send notification to defined users
+        for user in users:
+            record.activity_schedule(
+                "edi_notification_oca.mail_activity_failed_exchange_record_warning",
```
Comment (simahawk) [inline edi_notification_oca/components/listener.py]: I would make the activity configurable on the type too w/ default to this one if available.

File: edi_notification_oca/components/listener.py [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+                summary=_("Error on process of record '%s'.", record.identifier),
+                note=record.exchange_error,
+                user_id=user.id,
```
Comment (simahawk) [inline edi_notification_oca/components/listener.py:36]:
```suggestion
                user_id=user.id,
                automated=True
```

File: edi_notification_oca/components/listener.py [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+        if not send_notification_on_error:
+            return True
+        if not exc_type.notification_groups_ids and not exc_type.notification_users_ids:
+            return True
+        users = self.env["res.users"]
+        if exc_type.notification_groups_ids:
+            for group in exc_type.notification_groups_ids:
+                users |= group.users
+        if exc_type.notification_users_ids:
+            users |= exc_type.notification_users_ids
```
Comment (simahawk) [inline edi_notification_oca/components/listener.py]:
```suggestion
        users = self._get_users_to_notify(record, exc_type)
```

File: edi_notification_oca/components/listener.py [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+        # Send notification to defined users
+        for user in users:
+            record.activity_schedule(
+                "edi_notification_oca.mail_activity_failed_exchange_record_warning",
+                summary=_("Error on process of record '%s'.", record.identifier),
```
Comment (simahawk) [inline edi_notification_oca/components/listener.py]:
```suggestion
                summary=_("EDI: process error on record '%(identifier)s'.", identifier=record.identifier),
```

File: edi_notification_oca/tests/test_edi_notification.py [hunk 13]
```diff
@@ -0,0 +1,199 @@
...
+        self.assertEqual(len(a_noti), 1)
+        self.assertEqual(
+            a_noti.summary,
+            f"Error on process of record '{self.record.identifier}'.",
+        )
+        self.assertIn(
+            "OOPS! Something went wrong :(",
+            a_noti.note,
+        )
+        b_noti = self.env["mail.activity"].search(
```
Comment (simahawk) [inline edi_notification_oca/tests/test_edi_notification.py]: I'd recommend using `RecordCapturer` from `odoo.tests.common` :wink:

**Review by simahawk — CHANGES_REQUESTED** [review]

File: edi_notification_oca/models/edi_exchange_type.py (hunk 2 above)
Comment (QuocDuong1306) [reply edi_notification_oca/models/edi_exchange_type.py]: Hello @simahawk , i think we can handle some cases: we write through code with groups and users but forgot to enable  `notify_on_process_error`, the `onchange` didn't work with case...

Comment (simahawk) [reply edi_notification_oca/models/edi_exchange_type.py]: then maybe better to use an inverse?

**Comment by QuocDuong1306** [comment]
Thank @simahawk , I updated

File: edi_notification_oca/models/edi_exchange_type.py [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+        comodel_name="res.users",
+        string="Notify Users On Process Error",
+        inverse="_inverse_notify_on_process_error_groups_users",
+        help="Select users to send notifications to. If 'Notification Groups' "
+        "have been selected, notifications will also be sent to users selected in here.",
+    )
+    notify_on_process_error_activity_type_id = fields.Many2one(
+        "mail.activity.type",
+        string="Activity Type Used When Notify On Process Error",
+        default=_default_notify_on_process_error_activity_type_id,
```
Comment (simahawk) [inline edi_notification_oca/models/edi_exchange_type.py]: Rule of thumb: always avoid referencing methods' objects. Use lambdas to proxy the call
```suggestion
        default=lambda self: self._default_notify_on_process_error_activity_type_id(),
```
and keep the default method after the fields def

File: edi_notification_oca/tests/test_edi_notification.py [hunk 15]
```diff
@@ -0,0 +1,176 @@
...
+            FakeInputProcess,
+        )
+        cls._load_module_components(cls, "edi_notification_oca")
+        vals = {
+            "model": cls.partner._name,
+            "res_id": cls.partner.id,
+            "exchange_file": base64.b64encode(b"1234"),
+        }
+        cls.record = cls.backend.create_record("test_csv_input", vals)
+        cls.group_sale = cls.env.ref("sales_team.group_sale_salesman")
```
Comment (simahawk) [inline edi_notification_oca/tests/test_edi_notification.py]: this implies a dependency on sales_team that is not necessary

**Review by simahawk — CHANGES_REQUESTED** [review]

**Comment by QuocDuong1306** [comment]
Thank @simahawk , I updated

**Comment by cyrilmanuel** [comment]
@simahawk is it ok for you ?

File: edi_notification_oca/__manifest__.py [hunk 16]
```diff
@@ -0,0 +1,17 @@
+# Copyright 2024 Camptocamp SA
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl).
+
+{
+    "name": "EDI Notification OCA",
+    "summary": """Define Notification Activities When Exchange Record.""",
```
Comment (simahawk) [inline edi_notification_oca/__manifest__.py]: when exchange record.... what?

File: edi_notification_oca/components/listener.py [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+
+    def _get_users_to_notify(self, exc_type):
+        exc_type.ensure_one()
+        users = self.env["res.users"]
+        if exc_type.notify_on_process_error_groups_ids:
+            for group in exc_type.notify_on_process_error_groups_ids:
+                users |= group.users
+        if exc_type.notify_on_process_error_users_ids:
+            users |= exc_type.notify_on_process_error_users_ids
+        return users
```
Comment (simahawk) [inline edi_notification_oca/components/listener.py]:
```suggestion
        return exc_type.notify_on_process_error_groups_ids.users | exc_type.notify_on_process_error_users_ids
```

**Review by simahawk — COMMENTED** [review]
@QuocDuong1306 just 2 minor remarks. The rest LG, thanks!

File: edi_notification_oca/__manifest__.py [hunk 18]
```diff
@@ -0,0 +1,17 @@
+# Copyright 2024 Camptocamp SA
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl).
+
+{
+    "name": "EDI Notification OCA",
```
Comment (simahawk) [inline edi_notification_oca/__manifest__.py]:
```suggestion
    "name": "EDI Notification",
```

**Review by simahawk — APPROVED** [review]

**Pull request merged**